File: DSA/Recursion/Recursion 6/cpp/PrintSubsequence.cpp
#include <bits/stdc++.h>
#include <iostream>

using namespace std;

void printSubsequence(int index, vector<int> &arr, int nums[], int n) {
    if (index == n) {
        for (auto item: arr) {
            cout << item << " ";
        }
        cout << "\n";
        return;
    }

    arr.push_back(nums[index]);
    printSubsequence(index + 1, arr, nums, n);
    arr.pop_back();
    printSubsequence(index + 1, arr, nums, n);
}

int main() {
    int nums[] = {3, 1, 2, 4};
    vector<int> arr;
    printSubsequence(0, arr, nums, 4);
}

File: DSA/Recursion/Recursion 6/python/PrintSubsequence.py
def PrintSub(index: int, arr: [], nums: [], n: int):
    if index == n:
        print(arr)
        return

    arr.append(nums[index])
    PrintSub(index + 1, arr, nums, n)
    arr.pop()
    PrintSub(index + 1, arr, nums, n)


if __name__ == '__main__':
    nums = [3, 1, 2]
    arr = []
    n = len(nums)
    PrintSub(0, arr, nums, 3)
